jsonschema/reference.go

```diff
@@ -0,0 +1,168 @@
+package jsonschema
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"path"
+	"path/filepath"
+	"reflect"
+	"strconv"
+	"strings"
+
+	"github.com/franela/goreq"
+)
+
+// jsonRef represents a JSON Reference source and targetRef
+type jsonRef struct {
+	Source []string
+	Target string
+}
+
+// Dereference parse JSON string and replaces all $ref with the referenced data.
+func Dereference(schemaPath string, input []byte) (schemaJSON, []byte, error) {
+	var dj schemaJSON
+	if !strings.Contains(string(input), "$ref") {
+		return dj, input, nil
+	}
+
+	var data interface{}
+	json.Unmarshal([]byte(input), &data)
+	refs, err := walkInterface(data, []string{}, []jsonRef{})
+	if err != nil {
+		return dj, input, fmt.Errorf("unable to walk interface %s: %v", schemaPath, err)
+	}
+
+	for _, ref := range refs {
+		if !strings.HasPrefix(ref.Target, "#") {
+			if ref.Source[0] == "allOf" {
+				dj.AllOf = append(dj.AllOf, schemaRef{Ref: ref.Target})
+			}
+			if ref.Source[0] == "oneOf" {
+				dj.OneOf = append(dj.OneOf, schemaRef{Ref: ref.Target})
+			}
+		}
+
+		top := data
+		for i, item := range ref.Source {
+			if i < len(ref.Source)-1 {
+				// assuming integer item is slice[index] instead of map[string]
+				if intKey, err := strconv.Atoi(item); err == nil {
+					top = top.([]interface{})[intKey]
+				} else {
+					top = top.(map[string]interface{})[item]
+				}
+			} else {
+				targetRef, err := buildReference(schemaPath, data, ref.Target)
+				if err != nil {
+					return dj, input, fmt.Errorf("unable to build reference from %s: %v", ref.Target, err)
+				}
+				targetKeys := reflect.ValueOf(targetRef).MapKeys()
```

Comment:
It is a bit funky to do reflection to get a length, I think better would be for buildReference to not return an empty interface.